packages/core/src/extension/extension-messenger.ts

```diff
@@ -0,0 +1,96 @@
+import { IframeEventListener, IframeEventPoster } from '../common/iframe';
+
+import { ExtensionNotFoundError } from './errors';
+import { Extension } from './extension';
+import { ExtensionCommandHandlers } from './extension-command-handler';
+import { ExtensionOriginEventMap, stringToExtensionCommand } from './extension-origin-event';
+import { HostOriginEvent } from './host-origin-event';
+
+export class ExtensionMessenger {
+    private _extensions: Extension[] | undefined;
+    private _listeners: { [extensionId: string]: IframeEventListener<ExtensionOriginEventMap> } =
+        {};
+    private _posters: { [extensionId: string]: IframeEventPoster<HostOriginEvent> } = {};
+
+    listen(
+        extensions: Extension[],
+        extensionId: string,
+        extensionCommandHandlers: ExtensionCommandHandlers,
+    ): void {
+        this._extensions = extensions;
+
+        const extension = this._getExtensionById(extensionId);
+
+        if (!this._listeners[extensionId]) {
+            this._listeners[extensionId] = new IframeEventListener(extension.url);
+        }
+
+        const listener = this._listeners[extensionId];
+
+        listener.listen();
+
+        Object.entries(extensionCommandHandlers).forEach(([command, handler]) => {
+            const extensionCommand = stringToExtensionCommand(command);
+
+            if (extensionCommand && handler) {
+                listener.addListener(extensionCommand, handler);
```

Comment:
If we take one handler instead and return unsubscribe method, it'd look something like this:

```ts
listener.addListener(extensionCommand, handler);

return () => listener.removeListener(extensionCommand, handler);
```

Reply:
I did follow this and add a unit test to cover it. Thanks.